File: anna/llm.js
// anna/llm.js
// ─────────────────────────────────────────────────────────────────────────────
// Provider-agnostic LLM client so Anna can run on Claude (Anthropic, default &
// recommended) OR another provider (OpenAI) — whichever key the team configures.
//
// Exposes one uniform shape regardless of provider:
//   llmComplete(...) -> { text, toolCalls: [{ name, input }], raw }
//   llmJSON(...)     -> first JSON object/array parsed from the reply
//
// Tool definitions are written once in Anthropic's `{ name, description,
// input_schema }` form and translated to OpenAI's function schema as needed.
// (Document/vision scanning stays Anthropic-only — see normalize.js.)
// ─────────────────────────────────────────────────────────────────────────────

const ANTHROPIC_URL = 'https://api.anthropic.com/v1/messages';
const OPENAI_URL = 'https://api.openai.com/v1/chat/completions';

export const PROVIDERS = ['anthropic', 'openai'];

export const DEFAULT_MODELS = {
  anthropic: { fast: process.env.ANNA_FAST_MODEL || 'claude-haiku-4-5-20251001', smart: process.env.ANNA_SMART_MODEL || 'claude-opus-4-8' },
  openai: { fast: process.env.ANNA_OPENAI_FAST_MODEL || 'gpt-4o-mini', smart: process.env.ANNA_OPENAI_SMART_MODEL || 'gpt-4o' },
};

export function providerModel(provider = 'anthropic', tier = 'fast', override) {
  if (override) return override;
  return (DEFAULT_MODELS[provider] || DEFAULT_MODELS.anthropic)[tier] || DEFAULT_MODELS.anthropic.fast;
}

/**
 * @param {Object} p
 * @param {string} [p.provider='anthropic']
 * @param {string} p.apiKey
 * @param {string} [p.model]
 * @param {string} [p.system]
 * @param {Array}  p.messages  [{ role:'user'|'assistant', content:string }]
 * @param {Array}  [p.tools]   Anthropic-style tool defs.
 * @param {number} [p.maxTokens=1024]
 * @returns {Promise<{ text, toolCalls, raw }>}
 */
export async function llmComplete({ provider = 'anthropic', apiKey, model, system, messages = [], tools, maxTokens = 1024 } = {}) {
  if (!apiKey) throw new Error('No API key configured');
  if (provider === 'openai') return callOpenAI({ apiKey, model: providerModel('openai', 'fast', model), system, messages, tools, maxTokens });
  return callAnthropic({ apiKey, model: providerModel('anthropic', 'fast', model), system, messages, tools, maxTokens });
}

export async function llmJSON(opts) {
  const { text } = await llmComplete(opts);
  const m = text.match(/\{[\s\S]*\}|\[[\s\S]*\]/);
  if (!m) throw new Error('LLM response contained no JSON');
  return JSON.parse(m[0]);
}

async function callAnthropic({ apiKey, model, system, messages, tools, maxTokens }) {
  const res = await fetch(ANTHROPIC_URL, {
    method: 'POST',
    headers: { 'x-api-key': apiKey, 'anthropic-version': '2023-06-01', 'content-type': 'application/json' },
    body: JSON.stringify({ model, max_tokens: maxTokens, ...(system ? { system } : {}), ...(tools ? { tools } : {}), messages }),
  });
  if (!res.ok) throw new Error(`Anthropic API error ${res.status}: ${(await res.text().catch(() => '')).slice(0, 200)}`);
  const raw = await res.json();
  const text = (raw.content || []).filter((b) => b.type === 'text').map((b) => b.text).join('\n');
  const toolCalls = (raw.content || []).filter((b) => b.type === 'tool_use').map((b) => ({ name: b.name, input: b.input || {} }));
  return { text, toolCalls, raw };
}

async function callOpenAI({ apiKey, model, system, messages, tools, maxTokens }) {
  const msgs = [];
  if (system) msgs.push({ role: 'system', content: system });
  for (const m of messages) msgs.push({ role: m.role === 'assistant' ? 'assistant' : 'user', content: String(m.content || '') });
  const body = { model, max_tokens: maxTokens, messages: msgs };
  if (tools) body.tools = tools.map((t) => ({ type: 'function', function: { name: t.name, description: t.description, parameters: t.input_schema } }));
  const res = await fetch(OPENAI_URL, {
    method: 'POST',
    headers: { authorization: `Bearer ${apiKey}`, 'content-type': 'application/json' },
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error(`OpenAI API error ${res.status}: ${(await res.text().catch(() => '')).slice(0, 200)}`);
  const raw = await res.json();
  const msg = raw.choices?.[0]?.message || {};
  const toolCalls = (msg.tool_calls || []).map((tc) => ({ name: tc.function?.name, input: safeParse(tc.function?.arguments) }));
  return { text: msg.content || '', toolCalls, raw };
}

const safeParse = (s) => { try { return JSON.parse(s || '{}'); } catch { return {}; } };

File: backend/app/anna/llm.py
"""Provider-agnostic LLM client so Anna can run on Claude (Anthropic, default &
recommended) OR OpenAI — whichever key the team configures in Settings → Anna AI.

Python port of anna/llm.js. Exposes one uniform shape regardless of provider:

    llm_complete(...) -> {"text", "toolCalls": [{"name", "input"}], "raw"}
    llm_json(...)     -> first JSON object/array parsed from the reply

Tool definitions are written once in Anthropic's ``{name, description,
input_schema}`` form and translated to OpenAI's function schema as needed.
(Document/vision scanning stays Anthropic-only — see normalize.py / claude.py.)
"""
import json
import os
import re

import httpx

ANTHROPIC_URL = "https://api.anthropic.com/v1/messages"
OPENAI_URL = "https://api.openai.com/v1/chat/completions"

PROVIDERS = ["anthropic", "openai"]

DEFAULT_MODELS = {
    "anthropic": {
        "fast": os.environ.get("ANNA_FAST_MODEL") or "claude-haiku-4-5-20251001",
        "smart": os.environ.get("ANNA_SMART_MODEL") or "claude-opus-4-8",
    },
    "openai": {
        "fast": os.environ.get("ANNA_OPENAI_FAST_MODEL") or "gpt-4o-mini",
        "smart": os.environ.get("ANNA_OPENAI_SMART_MODEL") or "gpt-4o",
    },
}

_JSON_RE = re.compile(r"\{[\s\S]*\}|\[[\s\S]*\]")


def provider_model(provider: str = "anthropic", tier: str = "fast", override: str | None = None) -> str:
    if override:
        return override
    models = DEFAULT_MODELS.get(provider) or DEFAULT_MODELS["anthropic"]
    return models.get(tier) or DEFAULT_MODELS["anthropic"]["fast"]


async def llm_complete(
    *,
    provider: str = "anthropic",
    api_key: str | None = None,
    model: str | None = None,
    system=None,
    messages: list | None = None,
    tools=None,
    max_tokens: int = 1024,
) -> dict:
    """Uniform completion across providers → {text, toolCalls, raw}."""
    if not api_key:
        raise RuntimeError("No API key configured")
    messages = messages or []
    if provider == "openai":
        return await _call_openai(
            api_key=api_key, model=provider_model("openai", "fast", model),
            system=system, messages=messages, tools=tools, max_tokens=max_tokens,
        )
    return await _call_anthropic(
        api_key=api_key, model=provider_model("anthropic", "fast", model),
        system=system, messages=messages, tools=tools, max_tokens=max_tokens,
    )


async def llm_json(**opts) -> object:
    out = await llm_complete(**opts)
    m = _JSON_RE.search(out["text"])
    if not m:
        raise RuntimeError("LLM response contained no JSON")
    return json.loads(m.group(0))


async def _call_anthropic(*, api_key, model, system, messages, tools, max_tokens) -> dict:
    body = {"model": model, "max_tokens": max_tokens, "messages": messages}
    if system:
        body["system"] = system
    if tools:
        body["tools"] = tools
    async with httpx.AsyncClient(timeout=120) as client:
        res = await client.post(
            ANTHROPIC_URL,
            headers={"x-api-key": api_key, "anthropic-version": "2023-06-01", "content-type": "application/json"},
            json=body,
        )
    if not res.is_success:
        raise RuntimeError(f"Anthropic API error {res.status_code}: {res.text[:200]}")
    raw = res.json()
    blocks = raw.get("content") or []
    text = "\n".join(b["text"] for b in blocks if b.get("type") == "text")
    tool_calls = [{"name": b.get("name"), "input": b.get("input") or {}} for b in blocks if b.get("type") == "tool_use"]
    return {"text": text, "toolCalls": tool_calls, "raw": raw}


async def _call_openai(*, api_key, model, system, messages, tools, max_tokens) -> dict:
    msgs = []
    if system:
        msgs.append({"role": "system", "content": system})
    for m in messages:
        role = "assistant" if m.get("role") == "assistant" else "user"
        msgs.append({"role": role, "content": str(m.get("content") or "")})
    body = {"model": model, "max_tokens": max_tokens, "messages": msgs}
    if tools:
        body["tools"] = [
            {"type": "function", "function": {"name": t["name"], "description": t.get("description"), "parameters": t.get("input_schema")}}
            for t in tools
        ]
    async with httpx.AsyncClient(timeout=120) as client:
        res = await client.post(
            OPENAI_URL,
            headers={"authorization": f"Bearer {api_key}", "content-type": "application/json"},
            json=body,
        )
    if not res.is_success:
        raise RuntimeError(f"OpenAI API error {res.status_code}: {res.text[:200]}")
    raw = res.json()
    msg = ((raw.get("choices") or [{}])[0]).get("message") or {}
    tool_calls = [
        {"name": (tc.get("function") or {}).get("name"), "input": _safe_parse((tc.get("function") or {}).get("arguments"))}
        for tc in (msg.get("tool_calls") or [])
    ]
    return {"text": msg.get("content") or "", "toolCalls": tool_calls, "raw": raw}


def _safe_parse(s) -> dict:
    try:
        return json.loads(s or "{}")
    except Exception:  # noqa: BLE001
        return {}
